### DISCUSSION TITLE
How to write logs in one log file when functions are in different modules

### DISCUSSION THREAD

**Discussion by MrCoco5921** [SpriteOvO/spdlog-rs#63] [Q&A]
in spdlog/examples/07_async_pool_sink.rs, firstly, we must construct a logger, that is Arc<Logger> type. Then we use `info!(logger: logger, "hello async_pool_sink");` to write this sentence "hello async_pool_sink" into the specific log file. However, I notice that there must be a logger param.

my code is like this:
```rust
fn foo(logger: Arc<Logger>) {
    info!(logger:logger,  "foo");
}

fn test(logger: Arc<Logger>) {
    error!(logger:logger, "test");
}

fn main() -> Result<(), spdlog::Error> {
    const LOG_FILE: &str = "tmp/async_file_sink.log";
    let path: PathBuf = env::current_exe().unwrap().parent().unwrap().join(LOG_FILE);
    let file_sink: Arc<FileSink> = Arc::new(FileSink::builder().path(path).truncate(true).build()?);
    let async_pool_sink: Arc<AsyncPoolSink> = Arc::new(AsyncPoolSink::builder().sink(file_sink).build().unwrap());
    let logger: Arc<Logger> = Arc::new(Logger::builder().sink(async_pool_sink).build()?);
    info!(logger:logger, "hello async_pool_sink");
    foo(logger.clone());
    test(logger.clone());
    Ok(())
}
```

there are two functions foo and test needing to output log, but I don't want to send logger parameter to these functions.
So I want to ask if there is any solution to realize this：

`fn foo() {
      info!(logger:logger,  "foo");
  }`

 or

 `fn foo() {
      info!("foo");
  }`

 and the logs output into the same log file. 
 Thanks.

**Comment by SpriteOvO** [comment]
You can call `spdlog::set_default_logger()` with the logger you constructed, then just a `fn foo() { info!("foo"); }` to use your custom logger.

**Reply by MrCoco5921** [reply]
Thanks for your reply. However, when I call `spdlog::set_default_logger()` with the logger I constructed, log file can be generated, but is empty.
my code is like this:
```
fn main() -> Result<(), spdlog::Error> {
    const LOG_FILE: &str = "tmp/async_file_sink.log";

    let path: PathBuf = env::current_exe().unwrap().parent().unwrap().join(LOG_FILE);
    let file_sink: Arc<FileSink> = Arc::new(FileSink::builder().path(path).truncate(true).build()?);

    let async_pool_sink: Arc<AsyncPoolSink> =
        Arc::new(AsyncPoolSink::builder().sink(file_sink).build().unwrap());

    let logger: Arc<Logger> = Arc::new(Logger::builder().sink(async_pool_sink).build()?);

    set_default_logger(logger);

    info!("hello async_pool_sink");

    Ok(())
}
```
this log "hello async_pool_sink" doesn't exist in log file.

**Reply by SpriteOvO** [reply]
The default `flush_level_filter` of a logger is `MoreSevereEqual(Level::Info)`, this means that logs will only be flushed to files if there are warnings and more severe logs, or if the buffer in memory is full. This is for performance considerations.

If you want each log to be flushed to files in real time and don't care about performance, call [`logger.set_flush_level_filter(LevelFilter::All)`](https://docs.rs/spdlog-rs/0.3.13/spdlog/struct.Logger.html#method.set_flush_level_filter) for your logger.

If you want logs to be flushed on a periodic basis with performance in mind, just call [`logger.set_flush_period`](https://docs.rs/spdlog-rs/0.3.13/spdlog/struct.Logger.html#method.set_flush_period) for your logger. Note that warnings and more severe logs will still be flushed in real time because the default value of `flush_level_filter` works at the same time, so you won't miss them.

**Reply by MrCoco5921** [reply]
I try this:
```
let logger: Arc<Logger> = Arc::new(Logger::builder().sink(async_pool_sink).build()?);
logger.set_flush_level_filter(LevelFilter::All);
set_default_logger(logger);
warn!("hello async_pool_sink");  // also try error!
```
but file is still empty.

**Reply by SpriteOvO** [reply]
I reproduced the problem, regarding what's happening here, my guess is that the `main` function returns and exits before the asynchronous thread receives the log in time after this log is sent to it.

I think this should be considered a bug, and I'll see if it's possible to fix it later. But for now, you can add these two lines to the end of the `main` function as a workaround.

```rust
std::thread::sleep(std::time::Duration::from_millis(100));
spdlog::default_logger().flush();
```

Thanks for your feedback.

**Reply by SpriteOvO** [reply]
I created a issue #64 to track the bug.

**Reply by MrCoco5921** [reply]
Thanks. It works after I add these two line into my code. However, in my work project, I think it is inconvenient that I must add these two lines to see the log, and at the same time, project may throw some exceptions before running these two lines. So I would appreciate it if you could fix this problem.

**Reply by SpriteOvO** [reply]
Yes, those two lines are definitely expected to be unnecessary, I'll try to fix the bug. On the other hand, if you use `FileSink` directly instead of wrapping it with `AsyncPoolSink`, you can avoid encountering this bug as well.

Let me know if you have any other questions :)
